src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,237 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Christopher Hoskin
+-/
+import algebra.group.pi
+import algebra.group_power.lemmas
+
+/-!
+# Centroid homomorphisms
+
+This file defines centroid homomorphisms.
+
+We use the `fun_like` design, so each type of morphisms has a companion typeclass which is meant to
+be satisfied by itself and all stricter types.
+
+## Types of morphisms
+
+* `centroid_hom`: Maps which preserve left and right multiplication.
+
+## Typeclasses
+
+* `centroid_hom_class`
+
+## Tags
+
+centroid
+-/
+
+open function
+
+variables {F α : Type*}
+
+/-- The type of centroid homomorphisms from `α` to `α`. -/
+structure centroid_hom (α : Type*) [non_unital_non_assoc_semiring α] extends α →+ α :=
+(map_mul_left' (a b : α) : to_fun (a * b) = a * to_fun b)
+(map_mul_right' (a b : α) : to_fun (a * b) = to_fun a * b)
```

Comment:
```suggestion
section

-- Making `centroid_hom` an old structure will allow the lemma `to_add_monoid_hom_eq_coe`
-- to be true by `rfl`. After upgrading to Lean 4, this should no longer be needed
-- because eta for structures should provide the same result. 
set_option old_structure_cmd true

/-- The type of centroid homomorphisms from `α` to `α`. -/
structure centroid_hom (α : Type*) [non_unital_non_assoc_semiring α] extends α →+ α :=
(map_mul_left' (a b : α) : to_fun (a * b) = a * to_fun b)
(map_mul_right' (a b : α) : to_fun (a * b) = to_fun a * b)

end
```